### PR TITLE
Add copy capabilities

### PR DESCRIPTION
**Pull request by capitancambio**
Including an s3 based copy for improving performance when dealing with data stored in aws

### CODE REVIEW

File: src/tentaclio/fs/copiers.py [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+__all__ = ["DefaultCopier"]
+
+
+class DefaultCopier:
+    """Copier for those schemas that don't have an specialised implementation."""
+
+    def copy(self, source: URL, dest: URL):
+        """Copy the contents of the source url into the dest url."""
+        with open(str(source), mode="rb") as reader, open(str(dest), mode="wb") as writer:
+            cast(Writer, writer).write(cast(Reader, reader).read())
```
Comment (capitancambio) [inline src/tentaclio/fs/copiers.py]: mypy seems to be happy from other modules with the return value of open but not inside tentaclio. Therefore the cast

File: src/tentaclio/fs/copier.py [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+    ...
+
+
+class _CopierRegistryHolder:
+    """Module level singleton."""
+
+    instance: ClassVar[CopierRegistry] = CopierRegistry()
+
+
+COPIER_REGISTRY = _CopierRegistryHolder().instance
```
Comment (ig248) [inline src/tentaclio/fs/copier.py]: :coffee:

File: src/tentaclio/fs/api.py [hunk 3]
```diff
@@ -39,6 +40,19 @@ def listdir(url: str) -> Iterable[str]:
...
+def copy(source: str, dest: str):
+    """Copy the contents of the origin URL into the dest url.
+
+    Depending on the type of resource this can be much more performant than
+    >>> with tio.open(source) as reader, tio.open(dest) as writer:
+            writer.write(reader.read())
+    """
+    source_auth = credentials.authenticate(source)
+    dest_auth = credentials.authenticate(dest)
+    copier = COPIER_REGISTRY.get_handler(source_auth.scheme + "+" + dest_auth.scheme)
```
Comment (ig248) [inline src/tentaclio/fs/api.py]: I think it's safer to use a tuple instead of "magical" string constructs

File: src/tentaclio/fs/copiers.py (hunk 1 above)
Comment (ig248) [reply src/tentaclio/fs/copiers.py]: odd

File: src/tentaclio/fs/copier.py [hunk 4]
```diff
@@ -24,7 +24,13 @@ def copy(self, source: URL, dest: URL):
 
 
 class CopierRegistry(URLHandlerRegistry[Copier]):
-    """Registry for scanners."""
+    """Registry for scanners.
+
+    The scheme expected by this registry is a compound one having
+    the origin as first element and the destination as second element.
+    i.e. s3+s3, file+s3, ...
```
Comment (ig248) [inline src/tentaclio/fs/copier.py]: as per comment above

File: src/tentaclio/__init__.py [hunk 5]
```diff
@@ -53,3 +53,5 @@
 # Db registry
 DB_REGISTRY.register("postgresql", PostgresClient)
 DB_REGISTRY.register("awsathena+rest", AthenaClient)
+
+COPIER_REGISTRY.register("s3+s3", S3Client("s3://"))
```
Comment (ig248) [inline src/tentaclio/__init__.py]: as per comment above

File: tests/functional/s3/test_fs.py [hunk 6]
```diff
@@ -27,3 +27,17 @@ def test_scandir(fixture_client, test_bucket):
         key = base + f"{folder}"
         assert key in entries_by_url
         assert entries_by_url[key].is_dir
+
+
+def test_copy(fixture_client, test_bucket):
+    source = f"s3://{test_bucket}/source.txt"
+    dest = f"s3://{test_bucket}/dest.txt"
+    with tio.open(source, mode="w") as f:
+        f.write("Ph'nglui mglw'nafh Cthulhu R'lyeh wgah'nagl fhtagn.")
```
Comment (ig248) [inline tests/functional/s3/test_fs.py]: :ocean: :octopus:

File: src/tentaclio/fs/api.py (hunk 3 above)
Comment (capitancambio) [reply src/tentaclio/fs/api.py]: it's not a _magical construct_ per se, all the registries in tentaclio are indexed by the url scheme, and having two schemes appended by a `+` is more than common. i.e. "git+ssh", "athena+rest" 
https://en.wikipedia.org/wiki/Uniform_Resource_Identifier#Generic_syntax
so it's a valid url and keeps the consistency across the library.

File: src/tentaclio/__init__.py (hunk 5 above)
Comment (capitancambio) [reply src/tentaclio/__init__.py]: check the line above

File: src/tentaclio/fs/api.py (hunk 3 above)
Comment (ig248) [reply src/tentaclio/fs/api.py]: fair enough

**Pull request merged**